bart/sched/pelt.py

```diff
@@ -0,0 +1,377 @@
+                     default is the task used to intialize the Simulator
+                     (when specified)
+        :type  task: Task
+
+        :param start_s: the start time in [s]
+        :type  start_s: float
+
+        :param end_s: the end time in [s]
+        :type  end_s: float
+        """
+
+        # Computed PELT samples
+        samples = []
+        # Decay capping support
+        running_prev = True
+        capping = None
+
+        # Validate input parameters
+        if not isinstance(task, Task):
+            raise ValueError("Wrong time for task parameter")
+
+        # Intervals (in PELT samples) for the signal to compute
+        start_us = int(1e6 * Simulator._sample_us/1e3 * start_s)
+        end_us   = int(1e6 * Simulator._sample_us/1e3 * end_s)
+        #print "Simulation Time Range : ({}:{}) [us]".format(start_us, end_us)
+
+        # Add initial value to the generated signal
+        util = self.init_util
+        t_us = start_us
+        samples.append((float(t_us)/1e6, 0, False, util))
+
+        # Compute following PELT samples
+        t_us = start_us + Simulator._sample_us
+        while t_us <= end_us:
+
+            # Check if the task was running in the current PELT sample
+            running = task.running(float(t_us/1e3))
+
+            # Keep track of sleep start and decay capping time
+            if self.decay_cap_ms and running_prev and not running:
+                capping = t_us + (self.decay_cap_ms * 1e3)
+                #print "Set capping @{}".format(t_capping)
+
+            # Assume the task was running for all the current PELT sample
+            active_us = Simulator._sample_us if running else 0
+
+            # Always update utilization:
+            # - when the task is running
+            # - when is not running and we are not capping the decay
+            if running or not self.decay_cap_ms:
+                util = self._geom_sum(util, active_us)
+
+            # Update decay only up to the capping point
+            elif capping and t_us < capping:
+                util = self._geom_sum(util, active_us)
+
+            # Keep tracking the decay capped value
+            #else:
+            #    print "{}: capped @{} (since {})".format(t_us, util, capping)
+
+            # Append PELT samples
+            samples.append((float(t_us)/1e6, t_us/Simulator._sample_us, running, util))
+
+            # Prepare for next sample computation
+            running_prev = running
+            t_us += Simulator._sample_us
+
+        # Create DataFrame from computed samples
+        self._df = _DataFrame(samples, columns=['Time', 'PELT_Interval', 'Running', 'util_avg'])
+        self._df.set_index('Time', inplace=True)
+
+        return self._df
+
+    def stats(self, task, stats_start_s=0, stats_end_s=None):
+        """
+        Compute the stats over a pre-computed PELT signal, considering only
+        the specified portion of the signal.
```

Comment:
Maybe just me but I don't like the taste of the "precomputed signal" thing, it seems like an unnecessary inflexibility/global state. Why not just take the DataFrame as a parameter rather than using `self._df`?